### PR TITLE
Feature/entry image fix

### PR DESCRIPTION
**Pull request by thenav56**
NOTE: Client using file API might break.

Command to migrate data to the new structure: `python3 manage.py bulk_migrate entry_images_v2`

Issues:
-  Entry Excerpt: Add field `lead_image` in API #596
- Lead Preview: Send lead image file ID #595 
- Entry Excerpt: Only allow ID of file instead of URL in image field #594 
- Entry Excerpt: Migrate old image data to new mechanism #598 
- Gallery file: Provide only files uploaded by user via API #597

### CODE REVIEW

File: apps/bulk_data_migration/entry_images_v2/migrate.py [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+
+
+def _get_file_from_file_url(entry, string):
+    fileid = parse_number(string.rstrip('/').split('/')[-1])
+    return fileid and File.objects.filter(id=fileid).first()
+    try:
+        fileid = parse_number(string.rstrip('/').split('/')[-1])
+        return fileid and File.objects.filter(id=fileid).first()
+    except IndexError:
+        pass
```
Comment (annshress) [inline apps/bulk_data_migration/entry_images_v2/migrate.py]: Maybe, we only need `try...except` block here.

**Review by k9845 — APPROVED** [review]

File: apps/lead/models.py [hunk 2]
```diff
@@ -226,7 +226,10 @@ def get_for(cls, user):
         )
 
     def get_assignee(self):
-        return self.assignee.first()
+        # To be used by prefetch_related (Also, there is only one assignee)
+        assignees = self.assignee.all()
+        if assignees:
+            return assignees[0]
```
Comment (annshress) [inline apps/lead/models.py:265]: `self.assignee.first()` will suffice

Comment (AdityaKhatri) [reply apps/lead/models.py:265]: Please read the note above

File: apps/bulk_data_migration/entry_images_v2/migrate.py [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+def _get_file_from_s3_url(entry, string):
+    file_path = '/'.join(string.split('?')[0].split('/')[4:])
+    # NOTE: For lead-preview generate gallery files
+    if file_path.startswith('lead-preview/'):
+        lead_preview = LeadPreviewImage.objects.filter(file=file_path).first()
+        if lead_preview and lead_preview.file.storage.exists(lead_preview.file):
+            return lead_preview.clone_as_deep_file(entry.created_by)
+        return
+    try:
+        file_path = '/'.join(string.split('?')[0].split('/')[4:])
```
Comment (annshress) [inline apps/bulk_data_migration/entry_images_v2/migrate.py]: Redundant. This try block should have been at the top instead.

File: apps/bulk_data_migration/entry_images_v2/migrate.py [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+
+def _get_file_from_s3_url(entry, string):
+    file_path = '/'.join(string.split('?')[0].split('/')[4:])
+    # NOTE: For lead-preview generate gallery files
+    if file_path.startswith('lead-preview/'):
+        lead_preview = LeadPreviewImage.objects.filter(file=file_path).first()
+        if lead_preview and lead_preview.file.storage.exists(lead_preview.file):
+            return lead_preview.clone_as_deep_file(entry.created_by)
+        return
+    try:
```
Comment (annshress) [inline apps/bulk_data_migration/entry_images_v2/migrate.py:40]: Just like `lead-preview`, an explicit `if file_path .starts_with(...)` would be concrete. But if you think its fine, please resolve.

File: apps/lead/models.py (hunk 2 above)
Comment (thenav56) [reply apps/lead/models.py:265]: Basically `self.assignee.first()` will not use prefetch cache.

File: apps/bulk_data_migration/entry_images_v2/migrate.py (hunk 1 above)
Comment (thenav56) [reply apps/bulk_data_migration/entry_images_v2/migrate.py]: Done

File: apps/bulk_data_migration/entry_images_v2/migrate.py (hunk 3 above)
Comment (thenav56) [reply apps/bulk_data_migration/entry_images_v2/migrate.py]: Done

**Review by annshress — APPROVED** [review]

**Review by k9845 — APPROVED** [review]

File: apps/bulk_data_migration/entry_images_v2/migrate.py [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+    except IndexError:
+        return
+    return fileid and File.objects.filter(id=fileid).first()
+
+
+def _get_file_from_s3_url(entry, string):
+    try:
+        file_path = '/'.join(string.split('?')[0].split('/')[4:])
+    except IndexError:
+        return
```
Comment (annshress) [inline apps/bulk_data_migration/entry_images_v2/migrate.py:34]: Is the use of `...[4:]` okay? Just to be safe, if things change, and we always get the `IndexError`.

**Review by annshress — APPROVED** [review]

File: apps/bulk_data_migration/entry_images_v2/migrate.py (hunk 5 above)
Comment (thenav56) [reply apps/bulk_data_migration/entry_images_v2/migrate.py:34]: According to the data currently in production, this should work without any issue. Also, there is image_raw with the original value which we can use to analyze and fix remaining if found.

**Pull request merged**